src/charm.py

```diff
@@ -450,6 +466,48 @@ def rotate_session_secret_key(self, event: ActionEvent):
             logger.warning(warning_msg)
             event.log(warning_msg)
 
+    # Ensure the host key is present.
+    def ensure_hostkey_secret_key(self):
+        if not self.unit.is_leader():
+            return
+        try:
+            self.model.get_secret(label=HOST_KEY_SECRET_LABEL)
+        except SecretNotFoundError:
+            self.app.add_secret(new_host_key(), label=HOST_KEY_SECRET_LABEL)
+
+    # set_host_key is the action setting the host key. It validates the key, sets it in the secret and restarts
+    # the workload.
+    def set_host_key(self, event: ActionEvent) -> None:
+        if not self.unit.is_leader():
+            event.log("Cannot update secret from non-leader unit")
+            event.fail("Run this action on the leader unit")
+            return
+        host_key_b64 = event.params.get("host-key")
+        host_key = b64decode(host_key_b64).decode("utf-8")
+
+        if not is_valid_private_key(host_key):
+            event.log("Invalid private key: \n" + host_key)
+            event.fail("Invalid private key")
+            return
+
+        try:
+            secret = self.model.get_secret(label=HOST_KEY_SECRET_LABEL)
+            secret.set_content({HOST_KEY_LOOKUP: host_key})
+        except SecretNotFoundError:
+            secret = self.app.add_secret({HOST_KEY_LOOKUP: host_key}, label=HOST_KEY_SECRET_LABEL)
+        # Force a refresh of the secret content to flush old data.
+        secret.get_content(refresh=True)
+
+        # Set the output of the action.
+        event.set_results({"secret-id": secret.id})
+        try:
+            self._update_workload(event)
+        except RuntimeError:
+            # This exception will be raised when trying to defer the action event.
+            warning_msg = "updating workload failed, JIMM units weren't restarted, they might not be ready"
```

Comment:
Wondering if we should add something to suggest the user doesn't need to do anything just yet.
```suggestion
            warning_msg = "updating workload failed, JIMM units weren't restarted, they might not be ready.\n No manual intervention needed."
```